### ISSUE TITLE
Installation size using Homebrew is unacceptable (~350 MBs)

### ISSUE THREAD

**Issue by mhmdanas** [nushell/nushell#3896]
Well, not a bug with Nu per se, and not even sure if this should be reported here or on homebrew-core, so tell me if this should be opened there instead.

So I just run `brew upgrade` and was greeted with this:
```console
$ brew upgrade
...
==> Pouring nushell--0.35.0.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/nushell/0.35.0: 27 files, 343.2MB
...
```

(The size was also similar when I upgraded to 0.34.0, but I didn't notice at the time yet)

I'm not sure if this is known or not, but I thought I'd bring attention to this, as I don't think a shell should be taking up such a large chunk of disk space.

**Comment by fdncred** [comment]
We decided to prioritize build speed over build size. If you want to build with a smaller size add this to the root Cargo.toml.
```toml
[profile.release]
opt-level = 'z' #Optimize for size
lto = true #Link Time Optimization
codegen-units = 1 #Reduce parallel codegen units
```
If you're on linux you could strip the symbols too to make it smaller. Be warned, this could take a long time to build depending on your hardware.

**Comment by aminya** [comment]
> We decided to prioritize build speed over build size. 

Does that mean the current packages are using the "debug binaries"?

**Comment by jntrnr** [comment]
The current published binaries are optimized for speed rather than for size.

**Comment by aminya** [comment]
> The current published binaries are optimized for speed rather than for size.

Thanks for the clarification! The comment was a bit confusing since I thought the binaries are built with 0 optimizations to speed up the build (usually used during development).

**Comment by jntrnr** [comment]
I'm trying to build now to optimize for size just to see what the size difference is. Another reason we changed is the LTO setting, which we found really increased both the compile times (from 15 mins to 45 mins+) and the amount of resources it took to compile.

Still, it's good to do the science and just see how much of an impact the different settings have.

**Comment by jntrnr** [comment]
With the settings @fdncred lists above:

```
────┬───────────────────────┬──────┬─────────┬────────────────
 #  │         name          │ type │  size   │    modified    
────┼───────────────────────┼──────┼─────────┼────────────────
  0 │ nu                    │ File │ 31.6 MB │ 15 minutes ago 
  1 │ nu_plugin_binaryview  │ File │ 13.0 MB │ 6 minutes ago  
  2 │ nu_plugin_chart_bar   │ File │ 13.0 MB │ 20 minutes ago 
  3 │ nu_plugin_chart_line  │ File │ 13.1 MB │ 11 minutes ago 
  4 │ nu_plugin_fetch       │ File │ 13.5 MB │ 7 minutes ago  
  5 │ nu_plugin_from_bson   │ File │ 12.1 MB │ 12 minutes ago 
  6 │ nu_plugin_from_mp4    │ File │ 12.2 MB │ 3 minutes ago  
  7 │ nu_plugin_from_sqlite │ File │ 13.2 MB │ 3 minutes ago  
  8 │ nu_plugin_inc         │ File │ 12.2 MB │ 20 minutes ago 
  9 │ nu_plugin_match       │ File │ 13.0 MB │ 11 minutes ago 
 10 │ nu_plugin_post        │ File │ 13.6 MB │ 15 minutes ago 
 11 │ nu_plugin_ps          │ File │ 12.3 MB │ 20 minutes ago 
 12 │ nu_plugin_query_json  │ File │ 12.2 MB │ 12 minutes ago 
 13 │ nu_plugin_s3          │ File │ 16.1 MB │ a minute ago   
 14 │ nu_plugin_selector    │ File │ 13.1 MB │ 14 minutes ago 
 15 │ nu_plugin_start       │ File │ 13.0 MB │ 7 minutes ago  
 16 │ nu_plugin_sys         │ File │ 12.3 MB │ 3 minutes ago  
 17 │ nu_plugin_textview    │ File │ 15.9 MB │ 15 minutes ago 
 18 │ nu_plugin_to_bson     │ File │ 12.1 MB │ 16 minutes ago 
 19 │ nu_plugin_to_sqlite   │ File │ 13.2 MB │ 2 minutes ago  
 20 │ nu_plugin_tree        │ File │ 12.8 MB │ 7 minutes ago  
 21 │ nu_plugin_xpath       │ File │ 12.5 MB │ 12 minutes ago 
 22 │ nu_pretty_hex         │ File │  1.4 MB │ 29 minutes ago 
────┴───────────────────────┴──────┴─────────┴────────────────
```
Compared to our current defaults:
```
────┬─────────────────────────────────────────────────┬──────┬─────────┬────────────
 #  │                      name                       │ type │  size   │  modified  
────┼─────────────────────────────────────────────────┼──────┼─────────┼────────────
  0 │ /home/jt/.cargo/bin/nu                          │ File │ 48.2 MB │ 2 days ago 
  1 │ /home/jt/.cargo/bin/nu_plugin_core_fetch        │ File │ 24.2 MB │ 2 days ago 
  2 │ /home/jt/.cargo/bin/nu_plugin_core_inc          │ File │ 21.1 MB │ 2 days ago 
  3 │ /home/jt/.cargo/bin/nu_plugin_core_match        │ File │ 22.1 MB │ 2 days ago 
  4 │ /home/jt/.cargo/bin/nu_plugin_core_post         │ File │ 24.3 MB │ 2 days ago 
  5 │ /home/jt/.cargo/bin/nu_plugin_core_ps           │ File │ 21.0 MB │ 2 days ago 
  6 │ /home/jt/.cargo/bin/nu_plugin_core_sys          │ File │ 21.1 MB │ 2 days ago 
  7 │ /home/jt/.cargo/bin/nu_plugin_core_textview     │ File │ 25.5 MB │ 2 days ago 
  8 │ /home/jt/.cargo/bin/nu_plugin_extra_binaryview  │ File │ 21.5 MB │ 2 days ago 
  9 │ /home/jt/.cargo/bin/nu_plugin_extra_chart_bar   │ File │ 22.0 MB │ 2 days ago 
 10 │ /home/jt/.cargo/bin/nu_plugin_extra_chart_line  │ File │ 22.0 MB │ 2 days ago 
 11 │ /home/jt/.cargo/bin/nu_plugin_extra_from_bson   │ File │ 20.8 MB │ 2 days ago 
 12 │ /home/jt/.cargo/bin/nu_plugin_extra_from_sqlite │ File │ 22.6 MB │ 2 days ago 
 13 │ /home/jt/.cargo/bin/nu_plugin_extra_query_json  │ File │ 20.9 MB │ 2 days ago 
 14 │ /home/jt/.cargo/bin/nu_plugin_extra_s3          │ File │ 27.1 MB │ 2 days ago 
 15 │ /home/jt/.cargo/bin/nu_plugin_extra_selector    │ File │ 22.1 MB │ 2 days ago 
 16 │ /home/jt/.cargo/bin/nu_plugin_extra_start       │ File │ 21.7 MB │ 2 days ago 
 17 │ /home/jt/.cargo/bin/nu_plugin_extra_to_bson     │ File │ 20.8 MB │ 2 days ago 
 18 │ /home/jt/.cargo/bin/nu_plugin_extra_to_sqlite   │ File │ 22.5 MB │ 2 days ago 
 19 │ /home/jt/.cargo/bin/nu_plugin_extra_tree        │ File │ 21.7 MB │ 2 days ago 
 20 │ /home/jt/.cargo/bin/nu_plugin_extra_xpath       │ File │ 21.2 MB │ 2 days ago 
────┴─────────────────────────────────────────────────┴──────┴─────────┴────────────
```

So it looks like the size almost doubles just from not having LTO and compile for size turned on (which used to be the defaults we used)

**Comment by aminya** [comment]
I don't think the issue is the build settings. For example, PowerShell is 250 MB on Windows and not all of it is compiled to native binaries. 

The solution here is to add a plugin system that downloads the plugins on demand. Something similar to PowerShell's `install-module`.

**Comment by fdncred** [comment]
On Windows:
Current size (optimized for speed): 311,430,656
Optimize for size: 162,006,528

**Comment by mhmdanas** [comment]
Thank you all for answering my question. Although I disagree with the decision, I respect it regardless.

**Comment by fdncred** [comment]
@mhmdanas fyi - we made a change recently to optimize for size again. although it may not be as small as it could be, it may be a nice compromise.
https://github.com/nushell/nushell/blob/4e2d3ceaaf8cf898c762fb4f67f3101aa46f1f6a/Cargo.toml#L143-L144

**Comment by mhmdanas** [comment]
That sounds good. In my (possibly irrelevant) opinion though, it would be nice if something more was done. I can understand the argument behind optimizing for speed, but does LTO for instance hinder that? (note that I know little about Rust in general, so I might have mistaken assumptions here)

**Comment by aminya** [comment]
I don't think it is a good idea to use `z`. `z` turns of the loop vectorization, which means no SIMD usage, which is a huge performance hit. You should at least use `s` if you want.

Turning on LTO, would probably also reduce the size.

I recommend that you build `nu` binary with all the optimizations, and instead build less used plugins and binaries for size.

_Originally posted by @aminya in https://github.com/nushell/nushell/issues/3924#issuecomment-911715752_

**Comment by aminya** [comment]
> it would be nice if something more was done.

As I said [here](https://github.com/nushell/nushell/issues/3896#issuecomment-893287347), the solution is to add a plugin system that downloads the plugins on demand. Something similar to PowerShell's `install-module`.

**Comment by fdncred** [comment]
@aminya ya, we'd like to add a plugin system. we've talked many times about it. if you're interesting in contributing such a thing feel free. we've recently moved `fetch` into the core commands so that such a thing could be built into nushell.

also, if you're interested in doing some cross platform tests that list compile time and resulting binary size, based on your recommendations, we'll take a look at it. you may very well be right about `s` and `lto`.

**Comment by jntrnr** [comment]
I'd like to keep this issue open as we're actively working to shrink the size of the installation. Recently, we merged in some of the plugins into Nu itself, which cut down on duplicate code in the installation binary and we're doing further investigations here.

**Comment by fdncred** [comment]
@mhmdanas have you seen the size of the 0.40.0 release?

**Comment by mhmdanas** [comment]
Well, that's certainly weird; on Homebrew I don't see the huge improvement mentioned in the blog:
```
==> Pouring nushell--0.40.0.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/nushell/0.40.0: 23 files, 329.9MB
```

**Comment by fdncred** [comment]
homebrew appears to be compiling and signing their own releases which is why you don't see the savings.

we can't really have it both ways, small file sizes (strip+upx) and small memory footprint (no strip, no upx). we have to choose one or the other.

If you want small file sizes you can:
1. download a release from our repo
2. clone and compile it on your own and run strip on the executables which will give some savings
3. clone and compile it on your own, run strip, run upx which will be like what is on our repo

if you want small memory footprint, you can:
1. clone the repo and compile it on your own without --features=extra

in the future we may separate plugins out of the release entirely and make them their own repo. i think the only other way to reduce raw file size is to reduce dependencies.

personally, i always want the entire package with all batteries included.

**Comment by eugenesvk** [comment]
> We decided to prioritize speed over size

and how much speed did you gain with such a sizeable increase in size?

> * download a release from our repo

that doesn't seem gain anything. Homebrew's install is ~350m, repo's unpacked size ~370m

> separate plugins out of the release entirely

Is there a convenient list of plugins with a brief description of how they are used so it's easier to tell what can be easily deleted? Checked the docs, but only found a link to the repo with the plugins' source code, not general description

**Comment by jntrnr** [comment]
This should hopefully get addressed by engine-q, the upcoming engine for nushell. Here are some current sizes for release builds:

```
> ls ./target/release/ | where type == File | where ($it.name | path parse | $in.extension == "")
───┬──────────────────────────────────┬──────┬─────────┬──────────────
 # │               name               │ type │  size   │   modified   
───┼──────────────────────────────────┼──────┼─────────┼──────────────
 0 │ target/release/engine-q          │ File │ 47.0 MB │ 14 hours ago 
 1 │ target/release/nu_plugin_example │ File │  4.1 MB │ 14 hours ago 
 2 │ target/release/nu_plugin_gstat   │ File │  4.6 MB │ 14 hours ago 
 3 │ target/release/nu_plugin_inc     │ File │  4.0 MB │ 14 hours ago 
───┴──────────────────────────────────┴──────┴─────────┴──────────────
```

**Comment by fdncred** [comment]
one other thing that I think will potentially make our releases smaller is that with cargo 1.58 (iirc) there is supposed to be stabilized built in symbol stripping.

**Comment by jntrnr** [comment]
Just checking in here again: the latest main (which uses the new engine) should be much smaller. Here are the current sizes:

```
>  ls ./target/release/nu* | where ($it.name | path parse | $in.extension == "")
╭───┬──────────────────────────────────┬──────┬──────────┬─────────────╮
│ # │               name               │ type │   size   │  modified   │
├───┼──────────────────────────────────┼──────┼──────────┼─────────────┤
│ 0 │ target/release/nu                │ file │ 49.8 MiB │ an hour ago │
│ 1 │ target/release/nu_plugin_example │ file │  3.9 MiB │ an hour ago │
│ 2 │ target/release/nu_plugin_gstat   │ file │  4.4 MiB │ an hour ago │
│ 3 │ target/release/nu_plugin_inc     │ file │  3.9 MiB │ an hour ago │
│ 4 │ target/release/nu_plugin_query   │ file │  5.9 MiB │ an hour ago │
╰───┴──────────────────────────────────┴──────┴──────────┴─────────────╯
```

**Comment by fdncred** [comment]
32.1 MB in Windows for version 0.59. I vote we close this issue.

**Comment by jntrnr** [comment]
Agreed - I believe this has been fixed with recent main (0.59 and later).
